Library/Homebrew/livecheck/skip_conditions.rb

```diff
@@ -0,0 +1,228 @@
+# typed: true
+# frozen_string_literal: true
+
+require "livecheck/livecheck"
+
+module Homebrew
+  module Livecheck
+    # The `Livecheck::SkipConditions` module primarily contains methods that
+    # check for various formula/cask conditions where a check should be skipped.
+    #
+    # @api private
+    module SkipConditions
+      extend T::Sig
+
+      module_function
+
+      sig do
+        params(
+          formula_or_cask: T.any(Formula, Cask::Cask),
+          livecheckable:   T::Boolean,
+          full_name:       T::Boolean,
+          verbose:         T::Boolean,
+        ).returns(T.nilable(Hash))
```

Comment:
Would it make the API nicer to not make this nilable?

Reply:
Assuming we would be returning an empty hash when a condition doesn't apply, it seems like the main difference is that we would check the return value using `present?`/`blank?` instead of `is_a?(Hash)`.

From some simple testing, returning an empty hash instead of `nil` is 2x slower and, of course, we would be allocating a hash in each of these methods every time they're executed (not just when the conditions match). Each of these conditions is checked for every formula/cask in a given livecheck run and we regularly do homebrew/core tap runs over the 5000+ formulae. When none of these conditions apply to a formula (which is common), that would be five empty hash allocations per formula or four per cask.

Maybe I'm misunderstanding what you're suggesting or overlooking notable benefits. If not, it seems like the detriments of returning an empty hash would outweigh the negligible benefit here.

Reply:
> Assuming we would be returning an empty hash when a condition doesn't apply, it seems like the main difference is that we would check the return value using `present?`/`blank?` instead of `is_a?(Hash)`.

This feels significantly more Ruby-like (not returning `nil` but not checking the type).

> From some simple testing, returning an empty hash instead of `nil` is 2x slower

This sort of performance testing isn't terribly useful for a tool that hits the network extensively. What % difference does it make to the overall runtime of the `brew livecheck` command? I'd be amazed if it's 0.01% and, given that, it's not worth the readability/non-idiomatic code overhead.

In general it's not worth micro-optimising like this unless a tool like `brew prof` indicates that it's going to make a method taking e.g. >10% of the runtime much faster.

Reply:
To satisfy my curiosity, I profiled these two implementations side by side, running only this code.

Across the 5000+ formulae in homebrew/core, the `nil` version took a total of `0.240250` seconds and the hash version took `0.284576` seconds. To be expected, returning a hash does involve more memory allocation, 13.42 MB for `nil` vs. 20.71 MB for empty hash, but none of the memory is retained.

Considering how long it takes to make even one network request (let alone thousands), I can appreciate that the performance difference here is beyond meaningless. I'm not fond of the unnecessary memory allocation but I can appreciate that idiomatic code takes precedence over this concern in this context.

I've updated this to return an empty hash instead of `nil` when a condition doesn't apply. Thanks for bearing with me as I worked through this for myself.

Reply:
> Considering how long it takes to make even one network request (let alone thousands), I can appreciate that the performance difference here is beyond meaningless. I'm not fond of the unnecessary memory allocation but I can appreciate that idiomatic code takes precedence over this concern in this context.

Thanks ❤️